### PR TITLE
Add ViTMatte

### LINKED ISSUE #25040: Add ViTMatte model
### Model description

ViTMatte is a recently released model for alpha matting on images i.e. background removal. 

The model accepts an input image and trimap (manually labelled grayscale image outlining the rough border of the foreground object) and predicts the alpha mate for each pixel. 

It introduces a series of small adaptations to the ViT architecture - selective global attention + window attention; adding convolutional blocks between transformers blocks - to reduce computational complexity and enhancing the high-frequency information passed through the network. 

At the time of publishing, ViTMatte showed SOTA performance on Distinctions-646 and strong performance (> Mattformer) on Composition-1K.  

### Open source status

- [X] The model implementation is available
- [X] The model weights are available

### Provide useful links for the implementation

Github: https://github.com/hustvl/ViTMatte
Paper: https://arxiv.org/pdf/2305.15272.pdf
Demo: https://colab.research.google.com/drive/1Dc2qoJueNZQyrTU19sIcrPyRDmvuMTF3?usp=sharing

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds the ViTMatte model, an elegant approach to image matting, entirely relying on the Vision Transformer backbone doing the heavy work, with a lightweight head on top.

Here's a Colab notebook showcasing inference: https://colab.research.google.com/drive/1pWTn3Iur-NR2xUIyDE31dBgng_hXjSsn?usp=sharing. 

The model leverages [VitDet](https://arxiv.org/abs/2203.16527) as backbone, hence this PR adds VitDet as a standalone model as well. It then leverages the AutoBackbone class to use this model as a backbone for image matting.

Fixes #25040.

### CODE REVIEW

File: src/transformers/image_utils.py [hunk 1]
```diff
@@ -162,9 +162,9 @@ def infer_channel_dimension_format(image: np.ndarray) -> ChannelDimension:
     else:
         raise ValueError(f"Unsupported number of image dimensions: {image.ndim}")
 
-    if image.shape[first_dim] in (1, 3):
+    if image.shape[first_dim] in (1, 3, 4):
         return ChannelDimension.FIRST
-    elif image.shape[last_dim] in (1, 3):
+    elif image.shape[last_dim] in (1, 3, 4):
```
Comment (NielsRogge) [inline src/transformers/image_utils.py:167]: @amyeroberts ViTMatte uses 4 input channels rather than the common 1 or 3. This made me make the updates here. However this causes some tests of tests/utils/test_image_utils.py to fail:

```
FAILED tests/utils/test_image_utils.py::UtilFunctionTester::test_get_channel_dimension_axis - AssertionError: 0 != 2
FAILED tests/utils/test_image_utils.py::UtilFunctionTester::test_infer_channel_dimension - AssertionError: <ChannelDimension.FIRST: 'channels_first'> != <ChannelDimension.LAST: 'channels_last'>
```
Can we keep this and update the tests accordingly, or should I revert the changes here?

Comment (amyeroberts) [reply src/transformers/image_utils.py:167]: Let's keep this as is for the moment whilst this PR is still open. I'm currently working on making this more flexible for the image processors, which will hopefully be merged in soon.

File: docs/source/en/_toctree.yml [hunk 2]
```diff
@@ -25,147 +25,146 @@
...
-      - local: tasks/language_modeling
-        title: Causal language modeling
-      - local: tasks/masked_language_modeling
-        title: Masked language modeling
-      - local: tasks/translation
-        title: Translation
-      - local: tasks/summarization
-        title: Summarization
-      - local: tasks/multiple_choice
-        title: Multiple choice
```
Comment (NielsRogge) [inline docs/source/en/_toctree.yml:44]: Not sure what happened here.. 

My ruff version is 0.0.259, black version is 23.1.0

Comment (sgugger) [reply docs/source/en/_toctree.yml:44]: Neither ruff or black will touch a yaml file. Nor sure what happened either but it needs to be reverted.

File: src/transformers/models/auto/image_processing_auto.py [hunk 3]
```diff
@@ -106,6 +106,7 @@
         ("vit_hybrid", "ViTHybridImageProcessor"),
         ("vit_mae", "ViTImageProcessor"),
         ("vit_msn", "ViTImageProcessor"),
+        ("vitmatte", "VitMatteImageProcessor"),
```
Comment (sgugger) [inline src/transformers/models/auto/image_processing_auto.py:109]: Would be nice to have a default for `vitedet` as well (even if it doesn't have its own image processor).

File: src/transformers/models/vitdet/__init__.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import TYPE_CHECKING
+
+from ...utils import (
+    OptionalDependencyNotAvailable,
+    _LazyModule,
+    is_torch_available,
+)
```
Comment (sgugger) [inline src/transformers/models/vitdet/__init__.py:20]: This fits in one line.

File: src/transformers/models/vitdet/configuration_vitdet.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            The size (resolution) of each image.
+        pretrain_image_size (`int`, *optional*, defaults to 224):
+            The size (resolution) of each image during pretraining.
+        patch_size (`int`, *optional*, defaults to 16):
+            The size (resolution) of each patch.
+        num_channels (`int`, *optional*, defaults to 3):
+            The number of input channels.
+        qkv_bias (`bool`, *optional*, defaults to `True`):
+            Whether to add a bias to the queries, keys and values.
+        drop_path_rate (`float`, *optional*, defaults to `0.0`):
```
Comment (sgugger) [inline src/transformers/models/vitdet/configuration_vitdet.py:68]:
```suggestion
        drop_path_rate (`float`, *optional*, defaults to 0.0):
```

File: src/transformers/models/vitdet/test.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import torch
```
Comment (sgugger) [inline src/transformers/models/vitdet/test.py:1]: Should be removed.

File: src/transformers/models/vitdet/test_convnext.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import torch
```
Comment (sgugger) [inline src/transformers/models/vitdet/test_convnext.py:1]: Should be removed.

File: src/transformers/models/vitdet/modeling_vitdet.py [hunk 8]
```diff
@@ -0,0 +1,879 @@
...
+        if self.use_relative_position_embeddings:
+            # initialize relative positional embeddings
+            self.rel_pos_h = nn.Parameter(torch.zeros(2 * input_size[0] - 1, head_dim))
+            self.rel_pos_w = nn.Parameter(torch.zeros(2 * input_size[1] - 1, head_dim))
+
+            # TODO define this in init_weights
+            # rel_pos_zero_init=True,
+            # if not rel_pos_zero_init:
+            #     nn.init.trunc_normal_(self.rel_pos_h, std=0.02)
+            #     nn.init.trunc_normal_(self.rel_pos_w, std=0.02)
```
Comment (sgugger) [inline src/transformers/models/vitdet/modeling_vitdet.py:256]: Needs to be addressed.

File: src/transformers/models/vitdet/modeling_vitdet.py [hunk 9]
```diff
@@ -0,0 +1,879 @@
...
+            padding=1,
+            bias=False,
+        )
+        self.norm2 = VitDetLayerNorm(bottleneck_channels)
+        self.act2 = ACT2FN[config.hidden_act]
+
+        self.conv3 = nn.Conv2d(bottleneck_channels, out_channels, 1, bias=False)
+        self.norm3 = VitDetLayerNorm(out_channels)
+
+        # TODO define this in init_weights
```
Comment (sgugger) [inline src/transformers/models/vitdet/modeling_vitdet.py:394]: Needs to be addressed.

File: src/transformers/models/vitdet/modeling_vitdet.py [hunk 10]
```diff
@@ -0,0 +1,879 @@
...
+            # `trunc_normal_cpu` not implemented in `half` issues
+            module.weight.data = nn.init.trunc_normal_(
+                module.weight.data.to(torch.float32), mean=0.0, std=self.config.initializer_range
+            ).to(module.weight.dtype)
+            if module.bias is not None:
+                module.bias.data.zero_()
+        elif isinstance(module, nn.LayerNorm):
+            module.bias.data.zero_()
+            module.weight.data.fill_(1.0)
+        # elif isinstance(module, VitDetEmbeddings):
```
Comment (sgugger) [inline src/transformers/models/vitdet/modeling_vitdet.py:663]: Needs to be properly added or removed.

File: src/transformers/models/vitmatte/configuration_vitmatte.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+    Args:
+        backbone_config (`PretrainedConfig` or `dict`, *optional*, defaults to `VitDetConfig()`):
+            The configuration of the backbone model.
+        hidden_size (`int`, *optional*, defaults to 384):
+            The number of input channels of the decoder.
+        initializer_range (`float`, *optional*, defaults to 0.02):
+            The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
+        convstream_out (`List[int]`, *optional*, defaults to `[48, 96, 192]`):
+            ...
```
Comment (sgugger) [inline src/transformers/models/vitmatte/configuration_vitmatte.py:49]: Needs to be filled (here and below).

File: src/transformers/models/vitmatte/image_processing_vitmatte.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+
+class VitMatteImageProcessor(BaseImageProcessor):
+    r"""
+    Constructs a ViTMatte image processor.
+
+    Args:
+        do_rescale (`bool`, *optional*, defaults to `True`):
+            Whether to rescale the image by the specified scale `rescale_factor`. Can be overridden by the `do_rescale`
+            parameter in the `preprocess` method.
+        rescale_factor (`int` or `float`, *optional*, defaults to `1/255`):
```
Comment (sgugger) [inline src/transformers/models/vitmatte/image_processing_vitmatte.py:46]:
```suggestion
        rescale_factor (`int` or `float`, *optional*, defaults to 1/255):
```

File: src/transformers/models/vitmatte/modeling_vitmatte.py [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+        self.matting_head = VitMatteHead(in_channels=config.fusion_out[-1])
+
+    def init_weights(self):
+        self.apply(self._init_weights)
+
+    def _init_weights(self, module):
+        if isinstance(module, nn.Conv2d):
+            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
+            if module.bias is not None:
+                module.bias.data.zero_()
```
Comment (sgugger) [inline src/transformers/models/vitmatte/modeling_vitmatte.py:215]: This should be in the `_init_weights` method of the pretrained model.

File: src/transformers/models/vitmatte/modeling_vitmatte.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+
+    def _init_weights(self, module):
+        if isinstance(module, VitMattePreTrainedModel):
+            module.backbone.init_weights()
+            module.decoder.init_weights()
+
+    def init_weights(self):
+        """Initialize the weights"""
+        self.backbone.init_weights()
+        self.decoder.init_weights()
```
Comment (sgugger) [inline src/transformers/models/vitmatte/modeling_vitmatte.py:82]: This shouldn't be overloaded like this. The code above is enough.

File: src/transformers/models/vitmatte/test.py [hunk 15]
```diff
@@ -0,0 +1,18 @@
+import requests
```
Comment (sgugger) [inline src/transformers/models/vitmatte/test.py:1]: Shouldn't be here.

File: tests/models/vitdet/test_modeling_vitdet.py [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+        self,
+        parent,
+        batch_size=13,
+        image_size=30,
+        patch_size=2,
+        num_channels=3,
+        is_training=True,
+        use_labels=True,
+        hidden_size=32,
+        num_hidden_layers=5,
```
Comment (sgugger) [inline tests/models/vitdet/test_modeling_vitdet.py:55]: SHouldn't be more than 2 now, cc @ydshieh

File: tests/models/vitmatte/test_modeling_vitmatte.py [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+        self,
+        parent,
+        batch_size=13,
+        image_size=32,
+        patch_size=16,
+        num_channels=4,
+        is_training=True,
+        use_labels=False,
+        hidden_size=2,
+        num_hidden_layers=6,
```
Comment (sgugger) [inline tests/models/vitmatte/test_modeling_vitmatte.py:58]: Same here.

**Review by sgugger — COMMENTED** [review]
Thanks for adding this. There seems to be a bit of work left in the model initializations and some of the docstrings.

File: src/transformers/image_utils.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/image_utils.py:167]: Looks like I can use #25462 for this purpose

File: src/transformers/models/auto/image_processing_auto.py (hunk 3 above)
Comment (NielsRogge) [reply src/transformers/models/auto/image_processing_auto.py:109]: VitDet was currently only added for `VitDetBackbone`. An image processor can theoretically be added once Mask R-CNN is added (#25348), which uses VitDet as backbone (and then use `MaskRCNNImageProcessor` as default image processor).

**Comment by amyeroberts** [comment]
Before I start reviewing - could you separate out the addition of VitDet and VitMatte? They should have their own respective PRs.
